File: backend/app/schemas/tools.py
"""Tools-related Pydantic schemas"""

from pydantic import BaseModel, Field
from typing import Any, Dict, List, Optional


class HeatmapProbabilityResponse(BaseModel):
    """Model probabilities for a game in the odds heatmap"""
    game_id: str
    home_win_prob: float
    away_win_prob: float
    spread_confidence: Optional[float] = None  # From ai_spread_pick (0-100)
    total_confidence: Optional[float] = None  # From ai_total_pick (0-100)
    has_cached_analysis: bool


class UpsetCandidateToolsResponse(BaseModel):
    """One upset candidate from GET /api/tools/upsets"""
    game_id: str
    start_time: str
    league: str
    home_team: str
    away_team: str
    underdog_side: str
    underdog_team: str
    underdog_ml: int
    implied_prob: float
    model_prob: float
    edge: float
    confidence: float

    # ROI trust boosters (optional; safe for staggered deploys)
    books_count: int = 0
    best_underdog_ml: Optional[int] = None
    median_underdog_ml: Optional[int] = None
    price_spread: Optional[int] = None
    worst_underdog_ml: Optional[int] = None
    flags: List[str] = Field(default_factory=list)  # e.g. ["stale_odds_suspected","low_books"]
    odds_quality: str = "bad"  # good | thin | bad

    market_disagreement: Optional[str] = None
    reasons: List[str] = Field(default_factory=list)


class UpsetFinderToolsAccess(BaseModel):
    """Access state for Upset Finder results."""

    can_view_candidates: bool
    # "login_required" | "premium_required" | None
    reason: Optional[str] = None


class UpsetFinderToolsMeta(BaseModel):
    """Honest scanning meta for Upset Finder."""

    games_scanned: int
    games_with_odds: int
    missing_odds: int
    games_scanned_capped: Optional[bool] = None


class UpsetFinderToolsResponse(BaseModel):
    """Response for GET /api/tools/upsets"""
    sport: str
    window_days: int
    min_edge: float
    max_results: int
    min_underdog_odds: int
    generated_at: str
    access: UpsetFinderToolsAccess
    candidates: List[UpsetCandidateToolsResponse]
    meta: UpsetFinderToolsMeta
    # Populated only when backend couldn't compute real meta/results.
    error: Optional[str] = None

File: frontend/lib/api/types/tools.ts
export interface HeatmapProbabilityResponse {
  game_id: string
  home_win_prob: number
  away_win_prob: number
  spread_confidence: number | null
  total_confidence: number | null
  has_cached_analysis: boolean
}

export interface UpsetCandidateTools {
  game_id: string
  start_time: string
  league: string
  home_team: string
  away_team: string
  underdog_side: string
  underdog_team: string
  underdog_ml: number
  implied_prob: number
  model_prob: number
  edge: number
  confidence: number

  // ROI trust boosters (optional for staggered deploys)
  books_count?: number
  best_underdog_ml?: number | null
  median_underdog_ml?: number | null
  price_spread?: number | null
  worst_underdog_ml?: number | null
  flags?: string[]
  odds_quality?: "good" | "thin" | "bad"

  market_disagreement?: string | null
  reasons: string[]
}

export interface UpsetFinderToolsMeta {
  games_scanned: number
  games_with_odds: number
  missing_odds: number
  games_scanned_capped?: boolean | null
}

export interface UpsetFinderToolsAccess {
  can_view_candidates: boolean
  reason: 'login_required' | 'premium_required' | null
}

export interface UpsetFinderToolsResponse {
  sport: string
  window_days: number
  min_edge: number
  max_results: number
  min_underdog_odds: number
  generated_at: string
  access: UpsetFinderToolsAccess
  candidates: UpsetCandidateTools[]
  meta: UpsetFinderToolsMeta
  error?: string | null
}
